{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import re\n",
    "\n",
    "from pysam import FastaFile\n",
    "import pandas as pd\n",
    "from cmapPy.pandasGEXpress.parse import parse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_ROOT = Path('/Users/liang/Box/MyCPTAC/CPTAC_data_collection_v1/')\n",
    "\n",
    "refseq_fasta_pth = DATA_ROOT / 'DCC/RefSeq_20160914/RefSeq.20160914_Human_ucsc_hg19_customProDBnr_mito_150contams.fasta.gz'\n",
    "\n",
    "phospho_pth = DATA_ROOT / 'CPTAC2/BRCA_prospective/prosp-brca-v5.3-phosphoproteome-ratio-norm-NArm.gct'\n",
    "acetyl_pth = DATA_ROOT / 'CPTAC2/BRCA_prospective/prosp-brca-v5.3-acetylome-ratio-norm-NArm.gct'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_fa = FastaFile(str(refseq_fasta_pth))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_peptide_start_end(peptide, protein_id, sites):\n",
    "    # Make sure the peptide is part of the protein sequence\n",
    "    protein_seq = protein_fa.fetch(protein_id)\n",
    "    if not peptide.upper() in protein_seq:\n",
    "        return [None, None, False, 'Peptide not found in protein']\n",
    "    \n",
    "    first_res = sites[0][0].lower()\n",
    "    first_res_loc = int(sites[0][1:])\n",
    "    first_res_ix = peptide.find(first_res) \n",
    "    \n",
    "    peptide_start = first_res_loc - first_res_ix\n",
    "    peptide_end = first_res_loc + len(peptide) - first_res_ix - 1\n",
    "    if protein_seq[peptide_start - 1:peptide_end] != peptide.upper():\n",
    "        return [None, None, False, 'Invalid location']\n",
    "    return [peptide_start, peptide_end, True, None]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['id.description', 'geneSymbol', 'numColumnsVMsiteObserved', 'bestScore',\n",
       "       'bestDeltaForwardReverseScore', 'Best_scoreVML',\n",
       "       'Best_numActualVMSites_sty', 'Best_numLocalizedVMsites_sty',\n",
       "       'variableSites', 'sequence', 'sequenceVML',\n",
       "       'accessionNumber_VMsites_numVMsitesPresent_numVMsitesLocalizedBest_earliestVMsiteAA_latestVMsiteAA',\n",
       "       'protein_mw', 'species', 'speciesMulti', 'orfCategory',\n",
       "       'accession_number', 'accession_numbers', 'protein_group_num',\n",
       "       'entry_name', 'GeneSymbol'],\n",
       "      dtype='object', name='rhd')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_phospho_gct = parse(str(phospho_pth))\n",
    "raw_phospho_gct.row_metadata_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>rhd</th>\n",
       "      <th>variableSites</th>\n",
       "      <th>sequence</th>\n",
       "      <th>accession_number</th>\n",
       "      <th>protein_group_num</th>\n",
       "      <th>GeneSymbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_001611.1_S18s _1_1_18_18</th>\n",
       "      <td>[S18s]</td>\n",
       "      <td>ELLLPNWQGsGSHGLTIAQR</td>\n",
       "      <td>NP_001611.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>AHNAK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_001611.1_S41s _1_1_41_41</th>\n",
       "      <td>[S41s]</td>\n",
       "      <td>DDGVFVQEVTQNsPAAR</td>\n",
       "      <td>NP_001611.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>AHNAK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_001611.1_S93s _1_1_93_93</th>\n",
       "      <td>[S93s]</td>\n",
       "      <td>KGDRsPEPGQTWTR</td>\n",
       "      <td>NP_001611.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>AHNAK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_001611.1_T99t _1_1_99_99</th>\n",
       "      <td>[T99t]</td>\n",
       "      <td>SPEPGQtWTR</td>\n",
       "      <td>NP_001611.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>AHNAK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_001611.1_S110s _1_1_110_110</th>\n",
       "      <td>[S110s]</td>\n",
       "      <td>EVFSSCSsEVVLSGDDEEYQR</td>\n",
       "      <td>NP_001611.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>AHNAK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_001611.1_S115s _1_1_115_115</th>\n",
       "      <td>[S115s]</td>\n",
       "      <td>EVFSSCSSEVVLsGDDEEYQR</td>\n",
       "      <td>NP_001611.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>AHNAK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_001611.1_S135s _1_1_135_135</th>\n",
       "      <td>[S135s]</td>\n",
       "      <td>LKsEDGVEGDLGETQSR</td>\n",
       "      <td>NP_001611.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>AHNAK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_001611.1_T158t _1_1_158_158</th>\n",
       "      <td>[T158t]</td>\n",
       "      <td>RVtAYTVDVTGR</td>\n",
       "      <td>NP_001611.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>AHNAK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_001611.1_S177s _1_1_177_177</th>\n",
       "      <td>[S177s]</td>\n",
       "      <td>DIDISsPEFK</td>\n",
       "      <td>NP_001611.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>AHNAK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_001611.1_S201s _1_1_201_201</th>\n",
       "      <td>[S201s]</td>\n",
       "      <td>HELTEISNVDVETQsGK</td>\n",
       "      <td>NP_001611.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>AHNAK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "rhd                            variableSites               sequence  \\\n",
       "rid                                                                   \n",
       "NP_001611.1_S18s _1_1_18_18           [S18s]   ELLLPNWQGsGSHGLTIAQR   \n",
       "NP_001611.1_S41s _1_1_41_41           [S41s]      DDGVFVQEVTQNsPAAR   \n",
       "NP_001611.1_S93s _1_1_93_93           [S93s]         KGDRsPEPGQTWTR   \n",
       "NP_001611.1_T99t _1_1_99_99           [T99t]             SPEPGQtWTR   \n",
       "NP_001611.1_S110s _1_1_110_110       [S110s]  EVFSSCSsEVVLSGDDEEYQR   \n",
       "NP_001611.1_S115s _1_1_115_115       [S115s]  EVFSSCSSEVVLsGDDEEYQR   \n",
       "NP_001611.1_S135s _1_1_135_135       [S135s]      LKsEDGVEGDLGETQSR   \n",
       "NP_001611.1_T158t _1_1_158_158       [T158t]           RVtAYTVDVTGR   \n",
       "NP_001611.1_S177s _1_1_177_177       [S177s]             DIDISsPEFK   \n",
       "NP_001611.1_S201s _1_1_201_201       [S201s]      HELTEISNVDVETQsGK   \n",
       "\n",
       "rhd                            accession_number  protein_group_num GeneSymbol  \n",
       "rid                                                                            \n",
       "NP_001611.1_S18s _1_1_18_18         NP_001611.1                1.1      AHNAK  \n",
       "NP_001611.1_S41s _1_1_41_41         NP_001611.1                1.1      AHNAK  \n",
       "NP_001611.1_S93s _1_1_93_93         NP_001611.1                1.1      AHNAK  \n",
       "NP_001611.1_T99t _1_1_99_99         NP_001611.1                1.1      AHNAK  \n",
       "NP_001611.1_S110s _1_1_110_110      NP_001611.1                1.1      AHNAK  \n",
       "NP_001611.1_S115s _1_1_115_115      NP_001611.1                1.1      AHNAK  \n",
       "NP_001611.1_S135s _1_1_135_135      NP_001611.1                1.1      AHNAK  \n",
       "NP_001611.1_T158t _1_1_158_158      NP_001611.1                1.1      AHNAK  \n",
       "NP_001611.1_S177s _1_1_177_177      NP_001611.1                1.1      AHNAK  \n",
       "NP_001611.1_S201s _1_1_201_201      NP_001611.1                1.1      AHNAK  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peptide_df = raw_phospho_gct.row_metadata_df.loc[\n",
    "    # Only Keep RefSeq accessions\n",
    "    ~raw_phospho_gct.row_metadata_df['accession_number'].str.startswith('smORF'),\n",
    "    ['variableSites', 'sequence', 'accession_number', 'protein_group_num', 'GeneSymbol']\n",
    "]\n",
    "peptide_df['variableSites'] = peptide_df['variableSites'].str.findall(r'[A-Z]\\d+[a-z]')\n",
    "peptide_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check what residues are modified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "S    38740\n",
       "T     6772\n",
       "M      797\n",
       "N      580\n",
       "Y      409\n",
       "Q      149\n",
       "C       62\n",
       "Name: variableSites, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peptide_df['variableSites'].explode().apply(lambda x: x[0]).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only STY that are phosphosites\n",
    "peptide_df['phosphosites'] = peptide_df['variableSites'].apply(\n",
    "    lambda l: [x[:-1] for x in l if x[0] in 'STY']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the peptide location\n",
    "peptide_loc_range_df = pd.DataFrame(\n",
    "    map(calc_peptide_start_end, \n",
    "        peptide_df.sequence, peptide_df.accession_number, peptide_df.phosphosites),\n",
    "    index=peptide_df.index,\n",
    "    columns=['peptide_start', 'peptide_end', 'peptide_loc_valid', 'peptide_loc_invalid_reason'],\n",
    ").astype({\n",
    "    'peptide_start': pd.Int64Dtype(),\n",
    "    'peptide_end': pd.Int64Dtype(),\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variableSites</th>\n",
       "      <th>sequence</th>\n",
       "      <th>accession_number</th>\n",
       "      <th>protein_group_num</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>phosphosites</th>\n",
       "      <th>peptide_start</th>\n",
       "      <th>peptide_end</th>\n",
       "      <th>peptide_loc_valid</th>\n",
       "      <th>peptide_loc_invalid_reason</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [variableSites, sequence, accession_number, protein_group_num, GeneSymbol, phosphosites, peptide_start, peptide_end, peptide_loc_valid, peptide_loc_invalid_reason]\n",
       "Index: []"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combine the location and peptide information\n",
    "peptide_with_loc_df = pd.concat([peptide_df, peptide_loc_range_df], axis=1)\n",
    "# List all the invalid peptides\n",
    "peptide_with_loc_df[~peptide_loc_range_df['peptide_loc_valid']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_df = peptide_with_loc_df.loc[\n",
    "    :,\n",
    "    ['phosphosites', 'peptide_start', 'peptide_end', 'peptide_loc_valid', 'peptide_loc_invalid_reason']\n",
    "].reset_index()\n",
    "out_df['phosphosites'] = out_df['phosphosites'].str.join(';')\n",
    "out_df.to_csv(\n",
    "    phospho_pth.with_suffix('.peptide_location.tsv.gz'),\n",
    "    sep='\\t',\n",
    "    index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rid</th>\n",
       "      <th>phosphosites</th>\n",
       "      <th>peptide_start</th>\n",
       "      <th>peptide_end</th>\n",
       "      <th>peptide_loc_valid</th>\n",
       "      <th>peptide_loc_invalid_reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_001611.1_S18s _1_1_18_18</td>\n",
       "      <td>S18</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_001611.1_S41s _1_1_41_41</td>\n",
       "      <td>S41</td>\n",
       "      <td>29</td>\n",
       "      <td>45</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_001611.1_S93s _1_1_93_93</td>\n",
       "      <td>S93</td>\n",
       "      <td>89</td>\n",
       "      <td>102</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_001611.1_T99t _1_1_99_99</td>\n",
       "      <td>T99</td>\n",
       "      <td>93</td>\n",
       "      <td>102</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP_001611.1_S110s _1_1_110_110</td>\n",
       "      <td>S110</td>\n",
       "      <td>103</td>\n",
       "      <td>123</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              rid phosphosites  peptide_start  peptide_end  \\\n",
       "0     NP_001611.1_S18s _1_1_18_18          S18              9           28   \n",
       "1     NP_001611.1_S41s _1_1_41_41          S41             29           45   \n",
       "2     NP_001611.1_S93s _1_1_93_93          S93             89          102   \n",
       "3     NP_001611.1_T99t _1_1_99_99          T99             93          102   \n",
       "4  NP_001611.1_S110s _1_1_110_110         S110            103          123   \n",
       "\n",
       "   peptide_loc_valid peptide_loc_invalid_reason  \n",
       "0               True                       None  \n",
       "1               True                       None  \n",
       "2               True                       None  \n",
       "3               True                       None  \n",
       "4               True                       None  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Acetylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['id.description', 'geneSymbol', 'numColumnsVMsiteObserved', 'bestScore',\n",
       "       'bestDeltaForwardReverseScore', 'Best_scoreVML', 'variableSites',\n",
       "       'sequence', 'sequenceVML',\n",
       "       'accessionNumber_VMsites_numVMsitesPresent_numVMsitesLocalizedBest_earliestVMsiteAA_latestVMsiteAA',\n",
       "       'protein_mw', 'species', 'speciesMulti', 'orfCategory',\n",
       "       'accession_number', 'accession_numbers', 'protein_group_num',\n",
       "       'entry_name', 'GeneSymbol'],\n",
       "      dtype='object', name='rhd')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_acetyl_gct = parse(str(acetyl_pth))\n",
    "raw_acetyl_gct.row_metadata_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>rhd</th>\n",
       "      <th>variableSites</th>\n",
       "      <th>sequence</th>\n",
       "      <th>accession_number</th>\n",
       "      <th>protein_group_num</th>\n",
       "      <th>GeneSymbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_000468.1_K28k _1_1_28_28</th>\n",
       "      <td>[K28k]</td>\n",
       "      <td>DAHkSEVAHR</td>\n",
       "      <td>NP_000468.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_000468.1_K36k _1_1_36_36</th>\n",
       "      <td>[K36k]</td>\n",
       "      <td>FkDLGEENFK</td>\n",
       "      <td>NP_000468.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_000468.1_K44k _1_1_44_44</th>\n",
       "      <td>[K44k]</td>\n",
       "      <td>FKDLGEENFk</td>\n",
       "      <td>NP_000468.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_000468.1_K75k _1_1_75_75</th>\n",
       "      <td>[K75k]</td>\n",
       "      <td>LVNEVTEFAkTCVADESAENCDK</td>\n",
       "      <td>NP_000468.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_000468.1_K88k _1_1_88_88</th>\n",
       "      <td>[K88k]</td>\n",
       "      <td>TCVADESAENCDkSLHTLFGDK</td>\n",
       "      <td>NP_000468.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_000468.1_K97k _1_1_97_97</th>\n",
       "      <td>[K97k]</td>\n",
       "      <td>SLHTLFGDkLCTVATLR</td>\n",
       "      <td>NP_000468.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_000468.1_K117k _1_1_117_117</th>\n",
       "      <td>[K117k]</td>\n",
       "      <td>ETYGEMADCCAkQEPERNECFLQHK</td>\n",
       "      <td>NP_000468.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_000468.1_K130k_1_1_130_130</th>\n",
       "      <td>[Q118q, K130k]</td>\n",
       "      <td>qEPERNECFLQHkDDNPNLPR</td>\n",
       "      <td>NP_000468.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_000468.1_K160k _1_1_160_160</th>\n",
       "      <td>[K160k]</td>\n",
       "      <td>LVRPEVDVMCTAFHDNEETFLkK</td>\n",
       "      <td>NP_000468.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_000468.1_K161k _1_1_161_161</th>\n",
       "      <td>[K161k]</td>\n",
       "      <td>LVRPEVDVMCTAFHDNEETFLKkYLYEIAR</td>\n",
       "      <td>NP_000468.1</td>\n",
       "      <td>1.1</td>\n",
       "      <td>ALB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "rhd                              variableSites  \\\n",
       "rid                                              \n",
       "NP_000468.1_K28k _1_1_28_28             [K28k]   \n",
       "NP_000468.1_K36k _1_1_36_36             [K36k]   \n",
       "NP_000468.1_K44k _1_1_44_44             [K44k]   \n",
       "NP_000468.1_K75k _1_1_75_75             [K75k]   \n",
       "NP_000468.1_K88k _1_1_88_88             [K88k]   \n",
       "NP_000468.1_K97k _1_1_97_97             [K97k]   \n",
       "NP_000468.1_K117k _1_1_117_117         [K117k]   \n",
       "NP_000468.1_K130k_1_1_130_130   [Q118q, K130k]   \n",
       "NP_000468.1_K160k _1_1_160_160         [K160k]   \n",
       "NP_000468.1_K161k _1_1_161_161         [K161k]   \n",
       "\n",
       "rhd                                                   sequence  \\\n",
       "rid                                                              \n",
       "NP_000468.1_K28k _1_1_28_28                         DAHkSEVAHR   \n",
       "NP_000468.1_K36k _1_1_36_36                         FkDLGEENFK   \n",
       "NP_000468.1_K44k _1_1_44_44                         FKDLGEENFk   \n",
       "NP_000468.1_K75k _1_1_75_75            LVNEVTEFAkTCVADESAENCDK   \n",
       "NP_000468.1_K88k _1_1_88_88             TCVADESAENCDkSLHTLFGDK   \n",
       "NP_000468.1_K97k _1_1_97_97                  SLHTLFGDkLCTVATLR   \n",
       "NP_000468.1_K117k _1_1_117_117       ETYGEMADCCAkQEPERNECFLQHK   \n",
       "NP_000468.1_K130k_1_1_130_130            qEPERNECFLQHkDDNPNLPR   \n",
       "NP_000468.1_K160k _1_1_160_160         LVRPEVDVMCTAFHDNEETFLkK   \n",
       "NP_000468.1_K161k _1_1_161_161  LVRPEVDVMCTAFHDNEETFLKkYLYEIAR   \n",
       "\n",
       "rhd                            accession_number  protein_group_num GeneSymbol  \n",
       "rid                                                                            \n",
       "NP_000468.1_K28k _1_1_28_28         NP_000468.1                1.1        ALB  \n",
       "NP_000468.1_K36k _1_1_36_36         NP_000468.1                1.1        ALB  \n",
       "NP_000468.1_K44k _1_1_44_44         NP_000468.1                1.1        ALB  \n",
       "NP_000468.1_K75k _1_1_75_75         NP_000468.1                1.1        ALB  \n",
       "NP_000468.1_K88k _1_1_88_88         NP_000468.1                1.1        ALB  \n",
       "NP_000468.1_K97k _1_1_97_97         NP_000468.1                1.1        ALB  \n",
       "NP_000468.1_K117k _1_1_117_117      NP_000468.1                1.1        ALB  \n",
       "NP_000468.1_K130k_1_1_130_130       NP_000468.1                1.1        ALB  \n",
       "NP_000468.1_K160k _1_1_160_160      NP_000468.1                1.1        ALB  \n",
       "NP_000468.1_K161k _1_1_161_161      NP_000468.1                1.1        ALB  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peptide_df = raw_acetyl_gct.row_metadata_df.loc[\n",
    "    # Only Keep RefSeq accessions\n",
    "    ~raw_acetyl_gct.row_metadata_df['accession_number'].str.startswith('smORF'),\n",
    "    ['variableSites', 'sequence', 'accession_number', 'protein_group_num', 'GeneSymbol']\n",
    "]\n",
    "peptide_df['variableSites'] = peptide_df['variableSites'].str.findall(r'[A-Z]\\d+[a-z]')\n",
    "peptide_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "K    10078\n",
       "M      356\n",
       "N       61\n",
       "Q       10\n",
       "Name: variableSites, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peptide_df['variableSites'].explode().apply(lambda x: x[0]).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only K that are acetylsites\n",
    "peptide_df['acetylsites'] = peptide_df['variableSites'].apply(\n",
    "    lambda l: [x[:-1] for x in l if x[0] in 'K']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the peptide location\n",
    "peptide_loc_range_df = pd.DataFrame(\n",
    "    map(calc_peptide_start_end, \n",
    "        peptide_df.sequence, peptide_df.accession_number, peptide_df.acetylsites),\n",
    "    index=peptide_df.index,\n",
    "    columns=['peptide_start', 'peptide_end', 'peptide_loc_valid', 'peptide_loc_invalid_reason'],\n",
    ").astype({\n",
    "    'peptide_start': pd.Int64Dtype(),\n",
    "    'peptide_end': pd.Int64Dtype(),\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variableSites</th>\n",
       "      <th>sequence</th>\n",
       "      <th>accession_number</th>\n",
       "      <th>protein_group_num</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>acetylsites</th>\n",
       "      <th>peptide_start</th>\n",
       "      <th>peptide_end</th>\n",
       "      <th>peptide_loc_valid</th>\n",
       "      <th>peptide_loc_invalid_reason</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [variableSites, sequence, accession_number, protein_group_num, GeneSymbol, acetylsites, peptide_start, peptide_end, peptide_loc_valid, peptide_loc_invalid_reason]\n",
       "Index: []"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combine the location and peptide information\n",
    "peptide_with_loc_df = pd.concat([peptide_df, peptide_loc_range_df], axis=1)\n",
    "# List all the invalid peptides\n",
    "peptide_with_loc_df[~peptide_loc_range_df['peptide_loc_valid']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_df = peptide_with_loc_df.loc[\n",
    "    :,\n",
    "    ['acetylsites', 'peptide_start', 'peptide_end', 'peptide_loc_valid', 'peptide_loc_invalid_reason']\n",
    "].reset_index()\n",
    "out_df['acetylsites'] = out_df['acetylsites'].str.join(';')\n",
    "out_df.to_csv(\n",
    "    acetyl_pth.with_suffix('.peptide_location.tsv.gz'),\n",
    "    sep='\\t',\n",
    "    index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rid</th>\n",
       "      <th>acetylsites</th>\n",
       "      <th>peptide_start</th>\n",
       "      <th>peptide_end</th>\n",
       "      <th>peptide_loc_valid</th>\n",
       "      <th>peptide_loc_invalid_reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_000468.1_K28k _1_1_28_28</td>\n",
       "      <td>K28</td>\n",
       "      <td>25</td>\n",
       "      <td>34</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_000468.1_K36k _1_1_36_36</td>\n",
       "      <td>K36</td>\n",
       "      <td>35</td>\n",
       "      <td>44</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_000468.1_K44k _1_1_44_44</td>\n",
       "      <td>K44</td>\n",
       "      <td>35</td>\n",
       "      <td>44</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_000468.1_K75k _1_1_75_75</td>\n",
       "      <td>K75</td>\n",
       "      <td>66</td>\n",
       "      <td>88</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP_000468.1_K88k _1_1_88_88</td>\n",
       "      <td>K88</td>\n",
       "      <td>76</td>\n",
       "      <td>97</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           rid acetylsites  peptide_start  peptide_end  \\\n",
       "0  NP_000468.1_K28k _1_1_28_28         K28             25           34   \n",
       "1  NP_000468.1_K36k _1_1_36_36         K36             35           44   \n",
       "2  NP_000468.1_K44k _1_1_44_44         K44             35           44   \n",
       "3  NP_000468.1_K75k _1_1_75_75         K75             66           88   \n",
       "4  NP_000468.1_K88k _1_1_88_88         K88             76           97   \n",
       "\n",
       "   peptide_loc_valid peptide_loc_invalid_reason  \n",
       "0               True                       None  \n",
       "1               True                       None  \n",
       "2               True                       None  \n",
       "3               True                       None  \n",
       "4               True                       None  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
