{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import re\n",
    "\n",
    "from pysam import FastaFile\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_ROOT = Path('/Users/liang/Box/MyCPTAC/CPTAC_proteome_v3.0/')\n",
    "\n",
    "refseq_fasta_pth = DATA_ROOT / 'DCC/RefSeq_20111201/20111201_RefSeq_Human_37-Mouse_37_Trypsin.renamed.fasta.gz'\n",
    "\n",
    "phospho_pth = DATA_ROOT / 'CPTAC2/TCGA_OV_retrospective/mmc3_phosphoproteome_ischemia_filtered_peptide_annotation.tsv.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_fa = FastaFile(str(refseq_fasta_pth))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_peptide_start_end(peptide, protein_id, sites):\n",
    "    # Make sure the peptide is part of the protein sequence\n",
    "    protein_seq = protein_fa.fetch(protein_id)\n",
    "    if not peptide.upper() in protein_seq:\n",
    "        return [None, None, False, 'Peptide not found in protein']\n",
    "    \n",
    "    first_res = sites[0][0].lower()\n",
    "    first_res_loc = int(sites[0][1:])\n",
    "    first_res_ix = peptide.find(first_res) \n",
    "    \n",
    "    peptide_start = first_res_loc - first_res_ix\n",
    "    peptide_end = first_res_loc + len(peptide) - first_res_ix - 1\n",
    "    if protein_seq[peptide_start - 1:peptide_end] != peptide.upper():\n",
    "        return [None, None, False, 'Invalid location']\n",
    "    return [peptide_start, peptide_end, True, None]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['refseq_prot_id', 'Peptide', 'Site'], dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peptide_df = pd.read_table(phospho_pth)\n",
    "peptide_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq_prot_id</th>\n",
       "      <th>original_peptide</th>\n",
       "      <th>Site</th>\n",
       "      <th>phosphosites</th>\n",
       "      <th>peptide</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_000022.3</td>\n",
       "      <td>K.SS*PAFGDRR.C</td>\n",
       "      <td>ALAD-S215s</td>\n",
       "      <td>[S215s]</td>\n",
       "      <td>SsPAFGDRR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_000022.3</td>\n",
       "      <td>R.DAAKS*SPAFGDRR.C</td>\n",
       "      <td>ALAD-S214s</td>\n",
       "      <td>[S214s]</td>\n",
       "      <td>DAAKsSPAFGDRR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_000022.3</td>\n",
       "      <td>R.DAAKSS*PAFGDR.R</td>\n",
       "      <td>ALAD-S215s</td>\n",
       "      <td>[S215s]</td>\n",
       "      <td>DAAKSsPAFGDR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_000022.3</td>\n",
       "      <td>R.DAAKSS*PAFGDRR.C</td>\n",
       "      <td>ALAD-S215s</td>\n",
       "      <td>[S215s]</td>\n",
       "      <td>DAAKSsPAFGDRR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP_000028.3</td>\n",
       "      <td>-.PY*SVGFR.E</td>\n",
       "      <td>ANK1-Y3y</td>\n",
       "      <td>[Y3y]</td>\n",
       "      <td>PySVGFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NP_000028.3</td>\n",
       "      <td>-.PYS*VGFR.E</td>\n",
       "      <td>ANK1-S4s</td>\n",
       "      <td>[S4s]</td>\n",
       "      <td>PYsVGFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NP_000028.3</td>\n",
       "      <td>K.NGY*TPLHIAAK.K</td>\n",
       "      <td>ANK1-Y603y</td>\n",
       "      <td>[Y603y]</td>\n",
       "      <td>NGyTPLHIAAK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NP_000028.3</td>\n",
       "      <td>K.NGYT*PLHIAAK.K</td>\n",
       "      <td>ANK1-T604t</td>\n",
       "      <td>[T604t]</td>\n",
       "      <td>NGYtPLHIAAK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NP_000028.3</td>\n",
       "      <td>K.RLGYIS*VTDVLK.V</td>\n",
       "      <td>ANK1-S781s</td>\n",
       "      <td>[S781s]</td>\n",
       "      <td>RLGYIsVTDVLK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NP_000028.3</td>\n",
       "      <td>R.IT*HSPTVSQVTER.S</td>\n",
       "      <td>ANK1-T1684t</td>\n",
       "      <td>[T1684t]</td>\n",
       "      <td>ItHSPTVSQVTER</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  refseq_prot_id    original_peptide         Site phosphosites        peptide\n",
       "0    NP_000022.3      K.SS*PAFGDRR.C   ALAD-S215s      [S215s]      SsPAFGDRR\n",
       "1    NP_000022.3  R.DAAKS*SPAFGDRR.C   ALAD-S214s      [S214s]  DAAKsSPAFGDRR\n",
       "2    NP_000022.3   R.DAAKSS*PAFGDR.R   ALAD-S215s      [S215s]   DAAKSsPAFGDR\n",
       "3    NP_000022.3  R.DAAKSS*PAFGDRR.C   ALAD-S215s      [S215s]  DAAKSsPAFGDRR\n",
       "4    NP_000028.3        -.PY*SVGFR.E     ANK1-Y3y        [Y3y]        PySVGFR\n",
       "5    NP_000028.3        -.PYS*VGFR.E     ANK1-S4s        [S4s]        PYsVGFR\n",
       "6    NP_000028.3    K.NGY*TPLHIAAK.K   ANK1-Y603y      [Y603y]    NGyTPLHIAAK\n",
       "7    NP_000028.3    K.NGYT*PLHIAAK.K   ANK1-T604t      [T604t]    NGYtPLHIAAK\n",
       "8    NP_000028.3   K.RLGYIS*VTDVLK.V   ANK1-S781s      [S781s]   RLGYIsVTDVLK\n",
       "9    NP_000028.3  R.IT*HSPTVSQVTER.S  ANK1-T1684t     [T1684t]  ItHSPTVSQVTER"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peptide_df = peptide_df.rename(columns={'Peptide': 'original_peptide'})\n",
    "peptide_df['phosphosites'] = peptide_df['Site'].str.split('-', 1, expand=True).iloc[:, 1].str.findall(r'[A-Z]\\d+[a-z]')\n",
    "peptide_df['peptide'] = peptide_df['original_peptide'].str.slice(start=2, stop=-2).str.replace(r'([STY])\\*', lambda m: m.group(1).lower())\n",
    "peptide_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check what residues are modified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "S    5875\n",
       "T    1572\n",
       "Y     465\n",
       "Name: phosphosites, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peptide_df['phosphosites'].explode().apply(lambda x: x[0]).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only STY that are phosphosites\n",
    "peptide_df['phosphosites'] = peptide_df['phosphosites'].apply(\n",
    "    lambda l: [x[:-1] for x in l if x[0] in 'STY']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the peptide location\n",
    "peptide_loc_range_df = pd.DataFrame(\n",
    "    map(calc_peptide_start_end, \n",
    "        peptide_df.peptide, peptide_df.refseq_prot_id, peptide_df.phosphosites),\n",
    "    index=peptide_df.index,\n",
    "    columns=['peptide_start', 'peptide_end', 'peptide_loc_valid', 'peptide_loc_invalid_reason'],\n",
    ").astype({\n",
    "    'peptide_start': pd.Int64Dtype(),\n",
    "    'peptide_end': pd.Int64Dtype(),\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq_prot_id</th>\n",
       "      <th>original_peptide</th>\n",
       "      <th>Site</th>\n",
       "      <th>phosphosites</th>\n",
       "      <th>peptide</th>\n",
       "      <th>peptide_start</th>\n",
       "      <th>peptide_end</th>\n",
       "      <th>peptide_loc_valid</th>\n",
       "      <th>peptide_loc_invalid_reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>NP_000319.1</td>\n",
       "      <td>K.RVPS*INQK.I</td>\n",
       "      <td>RPGR-S961s</td>\n",
       "      <td>[S961]</td>\n",
       "      <td>RVPsINQK</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>Invalid location</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>NP_000356.1</td>\n",
       "      <td>K.TAT*PQQAQEVHEK.L</td>\n",
       "      <td>TPI1-T215t</td>\n",
       "      <td>[T215]</td>\n",
       "      <td>TAtPQQAQEVHEK</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>Invalid location</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>NP_000356.1</td>\n",
       "      <td>R.IIY*GGSVTGATCK.E</td>\n",
       "      <td>TPI1-Y246y</td>\n",
       "      <td>[Y246]</td>\n",
       "      <td>IIyGGSVTGATCK</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>Invalid location</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>NP_000356.1</td>\n",
       "      <td>R.IIYGGS*VTGATCK.E</td>\n",
       "      <td>TPI1-S249s</td>\n",
       "      <td>[S249]</td>\n",
       "      <td>IIYGGsVTGATCK</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>Invalid location</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>NP_000356.1</td>\n",
       "      <td>R.KQS*LGELIGTLNAAK.V</td>\n",
       "      <td>TPI1-S58s</td>\n",
       "      <td>[S58]</td>\n",
       "      <td>KQsLGELIGTLNAAK</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>Invalid location</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7764</th>\n",
       "      <td>NP_997657.1</td>\n",
       "      <td>K.SPPREGS*QGELTPANSQSR.M</td>\n",
       "      <td>GPS1-S474s</td>\n",
       "      <td>[S474]</td>\n",
       "      <td>SPPREGsQGELTPANSQSR</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>Invalid location</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7765</th>\n",
       "      <td>NP_997657.1</td>\n",
       "      <td>K.SPPREGSQGELT*PANSQSR.M</td>\n",
       "      <td>GPS1-T479t</td>\n",
       "      <td>[T479]</td>\n",
       "      <td>SPPREGSQGELtPANSQSR</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>Invalid location</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7766</th>\n",
       "      <td>NP_997657.1</td>\n",
       "      <td>R.GERDS*QTQAILTK.L</td>\n",
       "      <td>GPS1-S240s</td>\n",
       "      <td>[S240]</td>\n",
       "      <td>GERDsQTQAILTK</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>Invalid location</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7767</th>\n",
       "      <td>NP_997657.1</td>\n",
       "      <td>R.GERDSQT*QAILTK.L</td>\n",
       "      <td>GPS1-T242t</td>\n",
       "      <td>[T242]</td>\n",
       "      <td>GERDSQtQAILTK</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>Invalid location</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7768</th>\n",
       "      <td>NP_997657.1</td>\n",
       "      <td>R.NQIHVKS*PPR.E</td>\n",
       "      <td>GPS1-S468s</td>\n",
       "      <td>[S468]</td>\n",
       "      <td>NQIHVKsPPR</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>Invalid location</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1173 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     refseq_prot_id          original_peptide        Site phosphosites  \\\n",
       "118     NP_000319.1             K.RVPS*INQK.I  RPGR-S961s       [S961]   \n",
       "136     NP_000356.1        K.TAT*PQQAQEVHEK.L  TPI1-T215t       [T215]   \n",
       "137     NP_000356.1        R.IIY*GGSVTGATCK.E  TPI1-Y246y       [Y246]   \n",
       "138     NP_000356.1        R.IIYGGS*VTGATCK.E  TPI1-S249s       [S249]   \n",
       "139     NP_000356.1      R.KQS*LGELIGTLNAAK.V   TPI1-S58s        [S58]   \n",
       "...             ...                       ...         ...          ...   \n",
       "7764    NP_997657.1  K.SPPREGS*QGELTPANSQSR.M  GPS1-S474s       [S474]   \n",
       "7765    NP_997657.1  K.SPPREGSQGELT*PANSQSR.M  GPS1-T479t       [T479]   \n",
       "7766    NP_997657.1        R.GERDS*QTQAILTK.L  GPS1-S240s       [S240]   \n",
       "7767    NP_997657.1        R.GERDSQT*QAILTK.L  GPS1-T242t       [T242]   \n",
       "7768    NP_997657.1           R.NQIHVKS*PPR.E  GPS1-S468s       [S468]   \n",
       "\n",
       "                  peptide  peptide_start  peptide_end  peptide_loc_valid  \\\n",
       "118              RVPsINQK           <NA>         <NA>              False   \n",
       "136         TAtPQQAQEVHEK           <NA>         <NA>              False   \n",
       "137         IIyGGSVTGATCK           <NA>         <NA>              False   \n",
       "138         IIYGGsVTGATCK           <NA>         <NA>              False   \n",
       "139       KQsLGELIGTLNAAK           <NA>         <NA>              False   \n",
       "...                   ...            ...          ...                ...   \n",
       "7764  SPPREGsQGELTPANSQSR           <NA>         <NA>              False   \n",
       "7765  SPPREGSQGELtPANSQSR           <NA>         <NA>              False   \n",
       "7766        GERDsQTQAILTK           <NA>         <NA>              False   \n",
       "7767        GERDSQtQAILTK           <NA>         <NA>              False   \n",
       "7768           NQIHVKsPPR           <NA>         <NA>              False   \n",
       "\n",
       "     peptide_loc_invalid_reason  \n",
       "118            Invalid location  \n",
       "136            Invalid location  \n",
       "137            Invalid location  \n",
       "138            Invalid location  \n",
       "139            Invalid location  \n",
       "...                         ...  \n",
       "7764           Invalid location  \n",
       "7765           Invalid location  \n",
       "7766           Invalid location  \n",
       "7767           Invalid location  \n",
       "7768           Invalid location  \n",
       "\n",
       "[1173 rows x 9 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combine the location and peptide information\n",
    "peptide_with_loc_df = pd.concat([peptide_df, peptide_loc_range_df], axis=1)\n",
    "# List all the invalid peptides\n",
    "peptide_with_loc_df[~peptide_loc_range_df['peptide_loc_valid']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_df = peptide_with_loc_df.loc[\n",
    "    :,\n",
    "    ['refseq_prot_id', 'original_peptide', \n",
    "     'peptide', 'phosphosites', \n",
    "     'peptide_start', 'peptide_end', 'peptide_loc_valid', 'peptide_loc_invalid_reason']\n",
    "]\n",
    "out_df['phosphosites'] = out_df['phosphosites'].str.join(';')\n",
    "out_df.to_csv(\n",
    "    Path(phospho_pth.parent, phospho_pth.stem).with_suffix('.peptide_location.tsv.gz'),\n",
    "    sep='\\t',\n",
    "    index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq_prot_id</th>\n",
       "      <th>original_peptide</th>\n",
       "      <th>peptide</th>\n",
       "      <th>phosphosites</th>\n",
       "      <th>peptide_start</th>\n",
       "      <th>peptide_end</th>\n",
       "      <th>peptide_loc_valid</th>\n",
       "      <th>peptide_loc_invalid_reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_000022.3</td>\n",
       "      <td>K.SS*PAFGDRR.C</td>\n",
       "      <td>SsPAFGDRR</td>\n",
       "      <td>S215</td>\n",
       "      <td>214</td>\n",
       "      <td>222</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_000022.3</td>\n",
       "      <td>R.DAAKS*SPAFGDRR.C</td>\n",
       "      <td>DAAKsSPAFGDRR</td>\n",
       "      <td>S214</td>\n",
       "      <td>210</td>\n",
       "      <td>222</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_000022.3</td>\n",
       "      <td>R.DAAKSS*PAFGDR.R</td>\n",
       "      <td>DAAKSsPAFGDR</td>\n",
       "      <td>S215</td>\n",
       "      <td>210</td>\n",
       "      <td>221</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_000022.3</td>\n",
       "      <td>R.DAAKSS*PAFGDRR.C</td>\n",
       "      <td>DAAKSsPAFGDRR</td>\n",
       "      <td>S215</td>\n",
       "      <td>210</td>\n",
       "      <td>222</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP_000028.3</td>\n",
       "      <td>-.PY*SVGFR.E</td>\n",
       "      <td>PySVGFR</td>\n",
       "      <td>Y3</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  refseq_prot_id    original_peptide        peptide phosphosites  \\\n",
       "0    NP_000022.3      K.SS*PAFGDRR.C      SsPAFGDRR         S215   \n",
       "1    NP_000022.3  R.DAAKS*SPAFGDRR.C  DAAKsSPAFGDRR         S214   \n",
       "2    NP_000022.3   R.DAAKSS*PAFGDR.R   DAAKSsPAFGDR         S215   \n",
       "3    NP_000022.3  R.DAAKSS*PAFGDRR.C  DAAKSsPAFGDRR         S215   \n",
       "4    NP_000028.3        -.PY*SVGFR.E        PySVGFR           Y3   \n",
       "\n",
       "   peptide_start  peptide_end  peptide_loc_valid peptide_loc_invalid_reason  \n",
       "0            214          222               True                       None  \n",
       "1            210          222               True                       None  \n",
       "2            210          221               True                       None  \n",
       "3            210          222               True                       None  \n",
       "4              2            8               True                       None  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
